### PR TITLE
Bugfix/2899 2900 wsdl fixes

### PR DESCRIPTION
**Pull request by davidnich**

### CODE REVIEW

File: qlib/WSDL.qm [hunk 1]
```diff
@@ -1065,7 +1099,19 @@ public class WSDL::XsdArrayType inherits WSDL::XsdAbstractType {
             }
 
             default: {
-                throw SOAP_SERIALIZATION_ERROR, sprintf("don't know how to handle type %y", name);
+                string snsp = nsc.getOutputNamespacePrefix(SOAP_ENCODING);
+                /*if (True) throw "X";*/ if (val.typeCode() == NT_HASH && val.size() == 1) {
```
Comment (sejvlond) [inline qlib/WSDL.qm]: remove the commented code

File: qlib/WSDL.qm [hunk 2]
```diff
@@ -3057,41 +3137,46 @@ public class WSDL::SoapBinding inherits WSDL::OperationalBinding {
...
-                    else if (mrh.mrh{id}{key}.typeCode() == NT_HASH) {
-                        auto href = substr(mrh.mrh{id}{key}."^attributes^".href, 1);
+        # issue #2902: get all multiref elements with an "id" tag
+        foreach hash<auto> hh in (body.pairIterator()) {
+            if (!exists hh.value."^attributes^".id) {
+                continue;
+            }
+            hash<auto> h = remove body{hh.key};
+            string id = remove h."^attributes^".id;
+            # FIXME: process root & type
```
Comment (sejvlond) [inline qlib/WSDL.qm]: is there an issue opened to address this FIXME?

File: qlib/WSDL.qm [hunk 3]
```diff
@@ -1495,16 +1565,27 @@ public class WSDL::XsdComplexType inherits WSDL::XsdAbstractType {
...
+
+                            auto sa = seq.element."^attributes^";
+                            if (sa.maxOccurs != "unbounded") {
+                                throw "WSDL_ERROR", sprintf("array elements must have maxOccurs = \"unbounded\"; array type %y has maxOccurs = %y",
+                                    aa.arrayType, sa.maxOccurs);
+                            }
+
+                            elementmap = parseElements(seq.element, unresolved, nsc);
+                            cx_type = XET_SEQUENCE;
+                            #throw "WSDL_ERROR", sprintf("seq: %y; elementmap: %N", seq, elementmap);
```
Comment (sejvlond) [inline qlib/WSDL.qm]: please remove the commented code, if not needed

File: qlib/WSDL.qm [hunk 4]
```diff
@@ -1495,16 +1565,27 @@ public class WSDL::XsdComplexType inherits WSDL::XsdAbstractType {
...
+                        delete d.restriction.attribute;
+
+                        # issue #2899: check for an array element
+                        if (d.restriction.sequence) {
+                            auto seq = d.restriction.sequence;
+                            WSDL::XsdBase::removeNS(\seq);
+
+                            auto sa = seq.element."^attributes^";
+                            if (sa.maxOccurs != "unbounded") {
+                                throw "WSDL_ERROR", sprintf("array elements must have maxOccurs = \"unbounded\"; array type %y has maxOccurs = %y",
```
Comment (sejvlond) [inline qlib/WSDL.qm]: why must array elements have unbounded as maxOccurs? thx

File: qlib/WSDL.qm [hunk 5]
```diff
@@ -1046,11 +1050,41 @@ public class WSDL::XsdBaseType inherits WSDL::XsdAbstractType {
...
+        arrayType =~ s/\[\]$//;
+        if (!nsc.doType(arrayType, \typeinfo, \elementType)) {
+            unresolved.add(self);
+        }
+        else {
+            finalized = True;
+        }
+
+        #if (t != "binary")
+        #    throw "XSD-ARRAYTYPE-ERROR", sprintf("don't know how to handle arrays of type %y", t);
```
Comment (sejvlond) [inline qlib/WSDL.qm]: remove the commented code

**Review by sejvlond — CHANGES_REQUESTED** [review]

File: qlib/WSDL.qm (hunk 2 above)
Comment (davidnich) [reply qlib/WSDL.qm]: yes, qorelanguage/qore#2902

**Comment by davidnich** [comment]
@sejvlond I believe all requested changes have been implemented; also Ahmed just confirmed that the fixes are OK

**Review by sejvlond — APPROVED** [review]

**Pull request merged**